refactor(debug-cmd): migrate ParsePopulateArgs to CmdArgParser

XueSongTap opened the pull request:
## Changes

This PR refactors the `ParsePopulateArgs` function in `debugcmd.cc` to use the project's standard `CmdArgParser` class rather than manual argument parsing. The change:

- Replaces manual index-based argument access with structured parsing
- Uses `CmdArgParser` methods to track argument position and validate input
- Maintains identical validation rules and error messages
- Makes the code more consistent with other command implementations

## Testing

The implementation has been tested using:
- `cmd_arg_parser_test`
- `server_family_test`
- `dragonfly_test`

All tests pass with no changes to behavior.

## Motivation

related: #4820

adiholden commented:
@XueSongTap Thank you for this contribution.
The PR looks good, only one comment I have. For simple types such as string etc we use explicit types so instead of using auto please use the type in your code. Thank you

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,45 +685,52 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
-    string str = absl::AsciiStrToUpper(ArgS(args, index));
-    if (str == "RAND") {
+  while (parser.HasNext()) {
+    auto flag = absl::AsciiStrToUpper(parser.Next());
+    if (flag == "RAND") {
```
I think MapNext or TryMapNext is a better option. You don't need else if in this case

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -739,29 +746,29 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
       };
 
-      auto start = parse_slot(ArgS(args, ++index));
+      auto start = parse_slot(parser.Next());
```
you can do auto [start, end] = parser.Next<string_view, string_view>();

BorysTheDev replied:
By the way instead of parse_slot you can write the better option parser.Next<FInt<0, 16383>, FInt<0, 16383>>();
In this case you get the result in the most efficient way.

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -739,29 +746,29 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
-    } else if (str == "EXPIRE") {
-      if (args.size() < index + 3) {
+    } else if (flag == "EXPIRE") {
+      if (!parser.HasAtLeast(2)) {
```
you can avoid this check and do parser.Next<uint32_t, uint32_t>() and then check error. it will be shorter

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,45 +685,52 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
         return nullopt;
       }
-      if (!absl::SimpleAtoi(ArgS(args, ++index), &options.elements)) {
+      if (!absl::SimpleAtoi(parser.Next(), &options.elements)) {
```
parser.Next<int>() is better than do convert manually

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,45 +685,52 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
   PopulateOptions options;
-  if (!absl::SimpleAtoi(ArgS(args, 1), &options.total_count)) {
+
+  auto total_count = parser.Next();
+  if (parser.HasError()) {
+    builder->SendError(kUintErr);
+    return nullopt;
+  }
+  if (!absl::SimpleAtoi(total_count, &options.total_count)) {
```
you don't need to check for parser errors every time. parser designed in the way that you can check errors in the end.

XueSongTap commented:
@adiholden @BorysTheDev Thank you for your review. I've made the requested changes:

1. Used explicit type for flag variable instead of auto
2. Removed redundant HasNext() checks and relied on the error handling mechanism
3. Use Next() instead of doing convert manually
4. 
I'm still not entirely confident about the MapNext implementation and the enum flag definitions. Could you please pay special attention to these parts to ensure they follow the team's best practices? Specifically, I want to confirm that:
- The enum flags are properly defined and used
- The MapNext approach is correctly implemented for parsing the command arguments

Thank you for your guidance!

romange commented:
@XueSongTap - `debug populate 100 foo 1000 slots 0 100` fails on check-fail:
```
F20250331 08:37:25.331976 827411 cmd_arg_parser.cc:38] Check failed: !error_.has_value() Parsing error occured but not checked
*** Check failure stack trace: ***
    @     0x5e61bc9bd2c3  google::LogMessage::Fail()
    @     0x5e61bc9bd209  google::LogMessage::SendToLog()
    @     0x5e61bc9bc9bc  google::LogMessage::Flush()
    @     0x5e61bc9c0ab4  google::LogMessageFatal::~LogMessageFatal()
    @     0x5e61bc78144d  facade::CmdArgParser::~CmdArgParser()
    @     0x5e61bbdd225a  dfly::DebugCmd::ParsePopulateArgs()
    @     0x5e61bbdd23d0  dfly::DebugCmd::Populate()
    @     0x5e61bbdd0627  dfly::DebugCmd::Run()
    @     0x5e61bc1909eb  dfly::ServerFamily::Debug()

```

@BorysTheDev  can help you with this

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,95 +687,57 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
+    }
+    if (flag == FLAG_SLOT) {
+      auto [start, end] = parser.Next<FInt<0, 16383>, FInt<0, 16383>>();
+      options.slot_range = cluster::SlotRange{static_cast<SlotId>(start), static_cast<SlotId>(end)};
```
As I remember, slotId is std::uint16_t. I think you don't need static_cast

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,95 +687,57 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
+    PopulateFlag flag = parser.MapNext("RAND", FLAG_RAND, "TYPE", FLAG_TYPE, "ELEMENTS",
+                                       FLAG_ELEMENTS, "SLOT", FLAG_SLOT, "EXPIRE", FLAG_EXPIRE);
+
+    if (flag == FLAG_RAND) {
```
please use SWITCH CASE instead of IF

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,95 +687,57 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
-        builder->SendError(kSyntaxErr);
-        return nullopt;
-      }
+      continue;
```
I think you don't need continue if you use SWITCH CASE

BorysTheDev commented:
> @XueSongTap - `debug populate 100 foo 1000 slots 0 100` fails on check-fail:
> 
> ```
> F20250331 08:37:25.331976 827411 cmd_arg_parser.cc:38] Check failed: !error_.has_value() Parsing error occured but not checked
> *** Check failure stack trace: ***
>     @     0x5e61bc9bd2c3  google::LogMessage::Fail()
>     @     0x5e61bc9bd209  google::LogMessage::SendToLog()
>     @     0x5e61bc9bc9bc  google::LogMessage::Flush()
>     @     0x5e61bc9c0ab4  google::LogMessageFatal::~LogMessageFatal()
>     @     0x5e61bc78144d  facade::CmdArgParser::~CmdArgParser()
>     @     0x5e61bbdd225a  dfly::DebugCmd::ParsePopulateArgs()
>     @     0x5e61bbdd23d0  dfly::DebugCmd::Populate()
>     @     0x5e61bbdd0627  dfly::DebugCmd::Run()
>     @     0x5e61bc1909eb  dfly::ServerFamily::Debug()
> ```
> 
> @BorysTheDev can help you with this

You need to check error before destroy CmdArgParser object

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,95 +687,57 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
     return nullopt;
   }
```
I think you can remove this check, because when you call next you will get error if you don't have the arguments

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,95 +687,57 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
-    options.prefix = ArgS(args, 2);
+  options.total_count = parser.Next<uint64_t>();
+
+  if (parser.HasNext()) {
+    options.prefix = parser.Next<string_view>();
```
Consider using NextOrDefault instead of parser.HasNext()

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -685,95 +687,57 @@ optional<DebugCmd::PopulateOptions> DebugCmd::ParsePopulateArgs(CmdArgList args,
-      builder->SendError(kUintErr);
-      return nullopt;
-    }
+  if (parser.HasNext()) {
```
Consider using NextOrDefault instead of parser.HasNext()

XueSongTap commented:
hi, @BorysTheDev thank you for your review. i've Improved error handling and code organization in ParsePopulateArgs . Please take another look at my commit, and comment if you have any question or concern.

BorysTheDev commented:
Hi @XueSongTap I see you misunderstood my comment regarding error processing from CmdArgParser
Let me give you an example: you have a command with 3 arguments: CMD int string enum.
To parse such a command, you need to make the next steps:
```
parser.Next<int>()
parser.Next<string>()
parser.MapNext(...)
if (parser.HasError()) {
    return rb->SendError(parser.Error()->MakeReply());
}
res = ...;
return res;
```
So you need only one check of parser.HasError() and parser.Error() after last usage of parser. You don't need to check every parameter is it OK or not, parser returns kSyntaxError or kInvalidIntErr so you need another error processing only if the error is different. Errors like "Invalid elements count" don't provide any useful information because there are several reasons where is the problem and your error can be wrong.

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+  options.prefix = parser.NextOrDefault<string_view>("");
+  options.val_size = parser.NextOrDefault<uint32_t>(100);
+
+  if (parser.HasError()) {
+    builder->SendError("Invalid value size");
+    parser.Error();
+    return nullopt;
```
we don't need it here

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+    PopulateFlag flag = parser.MapNext("RAND", FLAG_RAND, "TYPE", FLAG_TYPE, "ELEMENTS",
+                                       FLAG_ELEMENTS, "SLOT", FLAG_SLOT, "EXPIRE", FLAG_EXPIRE);
+    if (parser.HasError()) {
+      builder->SendError("Unknown or invalid flag");
+      parser.Error();
```
just break

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+        break;
+      case FLAG_TYPE:
+        options.type = absl::AsciiStrToUpper(parser.Next<string_view>());
+        if (parser.HasError()) {
+          builder->SendError("Invalid type name");
+          parser.Error();
+          return nullopt;
```
just brake

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+        break;
+      case FLAG_ELEMENTS:
+        options.elements = parser.Next<uint32_t>();
+        if (parser.HasError()) {
+          builder->SendError("Invalid elements count");
+          parser.Error();
+          return nullopt;
```
break

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+        break;
+      case FLAG_SLOT: {
+        auto [start, end] = parser.Next<FInt<0, 16383>, FInt<0, 16383>>();
+        if (parser.HasError()) {
+          builder->SendError("Invalid slot range");
+          parser.Error();
+          return nullopt;
+        }
```
break

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
-        return nullopt;
+      case FLAG_EXPIRE: {
+        auto [min_ttl, max_ttl] = parser.Next<uint32_t, uint32_t>();
+        if (parser.HasError()) {
+          builder->SendError("Invalid expire range");
+          parser.Error();
+          return nullopt;
+        }
```
break

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+          parser.Error();
+          return nullopt;
+        }
+        if (min_ttl >= max_ttl) {
+          builder->SendError(kExpiryOutOfRange);
+          parser.Error();
+          return nullopt;
+        }
```
this is correct

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+        break;
       }
-      if (!absl::SimpleAtoi(ArgS(args, ++index), &end)) {
+      default:
```
This section is impossible so you can add  LOG(ERROR) here

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,96 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
-      if (!absl::SimpleAtoi(ArgS(args, ++index), &end)) {
+      default:
         builder->SendError(kSyntaxErr);
+        parser.Error();
```
you have this code after loop

XueSongTap commented:
@BorysTheDev Thank you for your feedback on error handling. I've simplified the code by implementing your suggestion to check for errors only once after completing all parsing operations. I've also:

1. Removed the individual error checks after each parameter parse
2. Added a single error check at the end of the function
3. Used parser.Error()->MakeReply() to get more accurate error information
4. Added LOG(ERROR) in the default case as it should be unreachable

Could you please review the changes and let me know if:
- The error handling structure is now correct
- The LOG(ERROR) message is appropriate
- There are any other issues that need to be addressed

Thank you!

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,58 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
       }
-      if (!absl::SimpleAtoi(ArgS(args, ++index), &end)) {
+      default:
+        LOG(ERROR) << "Error flag value in Populate arguments";
```
This section is impossible, so you don't need SendError and return from it. But you need to add info regarding cmd, so please add args print.

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,58 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+        auto [min_ttl, max_ttl] = parser.Next<uint32_t, uint32_t>();
+        if (min_ttl >= max_ttl) {
+          builder->SendError(kExpiryOutOfRange);
+          parser.Error();
```
(void)parser.Error(); to show that we don't care

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,58 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
+        break;
+      case FLAG_SLOT: {
+        auto [start, end] = parser.Next<FInt<0, 16383>, FInt<0, 16383>>();
+        options.slot_range = cluster::SlotRange{start, end};
```
cluster::SlotRange{SlotId(start), SlotId(end)}

XueSongTap replied:
Added explicit SlotId casting for slot range values

BorysTheDev commented:
@XueSongTap you forgot to address previous 2 comments:

- (void)parser.Error(); to show that we don't care
- This section is impossible, so you don't need SendError and return from it. But you need to add info regarding cmd, so please add args print.

XueSongTap commented on src/server/debugcmd.cc:
```diff
@@ -675,106 +675,58 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
-        return nullopt;
+  while (parser.HasNext()) {
+    PopulateFlag flag = parser.MapNext("RAND", FLAG_RAND, "TYPE", FLAG_TYPE, "ELEMENTS",
+                                       FLAG_ELEMENTS, "SLOT", FLAG_SLOT, "EXPIRE", FLAG_EXPIRE);
```
fix "SLOTS"

XueSongTap commented:
Hi @BorysTheDev, sorry for the oversight. In this update, I’ve improved the `(void)parser.Error();` to explicitly indicate that we don’t care about certain errors in this section and have args print added

Please let me know if you have any suggestions or if there’s anything more I should include in the error logging.

BorysTheDev commented:
@XueSongTap Could you rebase your branch, I don't understand why tests are failed

XueSongTap commented:
@BorysTheDev 
https://github.com/dragonflydb/dragonfly/actions/runs/14597395138/job/40965702767
```
64: [       OK ] RdbTest.SetExpiry (38 ms)
64: [ RUN      ] RdbTest.SaveFlush
64: I20250422 19:49:14.964385 17042 proactor_pool.cc:149] Running 3 io threads
64: I20250422 19:49:14.971694 17042 snapshot_storage.cc:182] Load snapshot: Searching for snapshot in directory: "/__w/dragonfly/dragonfly/build/src/server"
64: W20250422 19:49:14.971796 17042 server_family.cc:931] Load snapshot: No snapshot found
64: I20250422 19:49:14.971989 17042 test_utils.cc:259] Starting SaveFlush
64: I20250422 19:49:19.9[14784](https://github.com/dragonflydb/dragonfly/actions/runs/14597395138/job/40965702767#step:8:14785) 17103 db_slice.cc:150] Can't grow, free_items 40385, obj_bytes: 81  mem_available: 3178985
64: W20250422 19:49:19.915102 17103 db_slice.cc:691] AddOrFind: InsertNew failed, budget: 3178992 reclaimed: 0 offset: -7
64: W20250422 19:49:19.915180 17103 main_service.cc:1368]  SET :241818 xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx failed with reason: Out of memory
64: I20250422 19:49:20.914817 17102 db_slice.cc:150] Can't grow, free_items 31603, obj_bytes: 81  mem_available: 1119809
64: W20250422 19:49:20.915155 17102 db_slice.cc:691] AddOrFind: InsertNew failed, budget: 1119816 reclaimed: 0 offset: -7
64: W20250422 19:49:20.915227 17102 main_service.cc:1368]  SET :403196 xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx failed with reason: Out of memory
64: W20250422 19:49:21.675961 17103 main_service.cc:1017] Out of memory, used 40451808 ,rss 77881344 ,limit 40000000
64: W20250422 19:49:21.676077 17103 main_service.cc:1017] Out of memory, used 40451808 ,rss 77881344 ,limit 40000000
64: W20250422 19:49:21.676124 17103 main_service.cc:1017] Out of memory, used 40451808 ,rss 77881344 ,limit 40000000
64: W20250422 19:49:21.676168 17103 main_service.cc:1017] Out of memory, used 40451808 ,rss 77881344 ,limit 40000000
```

BorysTheDev commented on src/server/debugcmd.cc:
```diff
@@ -712,106 +712,57 @@ void DebugCmd::Migration(CmdArgList args, facade::SinkReplyBuilder* builder) {
-  }
+  options.total_count = parser.Next<uint64_t>();
+  options.prefix = parser.NextOrDefault<string_view>("");
+  options.val_size = parser.NextOrDefault<uint32_t>(100);
```
or you even can write `options.val_size = parser.NextOrDefault<uint32_t>(options.val_size);`

XueSongTap commented:
Hi, @BorysTheDev  Apologies for the incorrect default values in my previous commit. I've made the corrections as per your suggestions:
- Changed options.prefix default from "" to the correct "key"
- Changed options.val_size default from 100 to the correct 16

Could you please review and approve the CI? Thank you!

BorysTheDev commented:
@XueSongTap Hi, you forgot to apply pre-commit hooks. you need to trim whitespaces

XueSongTap commented:
Hi @BorysTheDev,  sorry for the oversight — I recently reconfigured my development environment and forgot to set up the pre-commit hooks. I've now amended the commit with the proper formatting, and everything should be in order. Thanks for your understanding!
